File: bindings/go/uhppote/uhppote.go
package uhppote

import (
    "net/netip"
    "os"
    "fmt"
)

type Controller struct {
    Controller uint32
    Address    string
    Transport  string
}

func GetAllControllers() ([]*GetControllerResponse, error) {
    request, err := GetControllerRequest(0)
    if err != nil {
        return nil, err
    }

    replies, err := broadcast(request)
    if err != nil {
        return nil, err
    }

    list := []*GetControllerResponse{}
    for _, reply := range replies {
        if response, err := getControllerResponse(reply); err != nil {
            return nil, err
        } else if response != nil {
            list = append(list, response)
        }
    }

    return list, nil
}

func Listen(events chan Event, errors chan error, interrupt chan os.Signal) error {
    ch := make(chan []uint8)

    go listen(ch)

loop:
    for {
        select {
        case msg := <-ch:
            if evt, err := event(msg); err != nil {
                errors <- err
            } else if evt != nil {
                events <- *evt
            }

        case <-interrupt:
            break loop
        }
    }

    return nil
}
{{- range .model.functions }}
{{ template "function" . -}}
{{end}}

{{define "function"}}
func {{CamelCase .name}}({{template "args" .args}}) {{if .response}}(*{{CamelCase .response.name}},error){{else}}error{{end}} {
    c := resolve(controller)

    {{if .response -}}
    request,err := {{CamelCase .request.name}}(c.Controller, {{template "params" slice .args 1}})
    if err != nil {
        return nil,err
    }

    if reply,err := send(c, request); err != nil {
        return nil,err
    } else if response,err := {{camelCase .response.name}}(reply); err != nil {
        return nil, err
    } else if response != nil {
        return response, nil
    }

    return nil, nil
    {{- else }}
    request, err := {{CamelCase .request.name}}(c.Controller, {{template "params" slice .args 1}})
    if err != nil {
        return err
    } 

    if _, err = send(c, request); err != nil {
        return err
    }
    
    return nil
    {{- end}}
}{{end}}

func resolve(controller any) Controller {
    switch v := controller.(type) {
        case uint32:
            address := ""
            transport := "udp"

            // if v == 405419896 {
            //     address = "192.168.1.100:60000"
            //     transport = "tcp"
            // }

            return Controller {
                Controller: v,
                Address: address,
                Transport: transport,
            }    

        case Controller:
            return v
    }

    panic(fmt.Sprintf("unknown controller type (%T)", controller))
}

File: bindings/rust/uhppoted/src/uhppote.rs
use std::net::Ipv4Addr;
use async_std::future;

use chrono::NaiveDateTime;
use chrono::NaiveDate;
use chrono::NaiveTime;

use decode::*;
use encode::*;
use ut0311::Msg;

use error::Error;

#[path = "encode.rs"]
pub mod encode;

#[path = "decode.rs"]
pub mod decode;

#[path = "ut0311.rs"]
mod ut0311;

#[path = "error.rs"]
pub mod error;

pub type PIN = u32;
pub type Event = decode::{{ CamelCase .model.event.name }};
pub type Result<T> = std::result::Result<T, Error>;
pub type Controller = ut0311::Controller;

#[allow(dead_code)]
trait IController {
    fn resolve(&self) -> ut0311::Controller;
}

{{range .model.responses}}
pub type {{CamelCase .name}} = decode::{{CamelCase .name}};
{{- end}}

pub fn set_bind_addr(addr: &str) {
    ut0311::set_bind_addr(addr)
}

pub fn set_broadcast_addr(addr: &str) {
    ut0311::set_broadcast_addr(addr)
}

pub fn set_listen_addr(addr: &str) {
    ut0311::set_listen_addr(addr)
}

pub fn set_debug(enabled: bool) {
    ut0311::set_debug(enabled)
}

pub async fn get_all_controllers() -> Result<Vec<GetControllerResponse>> {
    let request = get_controller_request(0)?;
    let response = ut0311::broadcast(&request).await;

    match response {
        Ok(replies) => {
            return Ok(replies
                .iter()
                .map(|reply| get_controller_response(&reply))
                .filter_map(|reply| reply.ok())
                .collect::<Vec<_>>());
        },
        Err(e) => return Err(e),
    }
}

pub fn listen(events: fn(Event), errors: fn(error::Error), interrupt: impl future::Future) -> Result<()> {
    let pipe = |msg: Msg| match decode::event(&msg) {
        Ok(event) => events(event),
        Err(e) => errors(e),
    };

    return ut0311::listen(pipe, errors, interrupt);
}

{{range .model.functions}}
{{- template "function" . -}}
{{end -}}

{{define "function"}}
pub async fn {{snakeCase .name}}({{template "args" .args}}) -> {{template "result" .}}{ 
    let request = {{snakeCase .request.name}}(controller.controller, {{template "params" slice .args 1}})?;
    let response = ut0311::send(controller, &request).await;

    {{if .response -}}
    match response {
        Ok(reply) =>  return Ok({{snakeCase .response.name}}(&reply)?),
        Err(e) => return Err(e),    
    }
    {{- else -}}
    match response {
        Ok(_) =>  return Ok(true),
        Err(e) => return Err(e),
    }
    {{end}}
}
{{end}}
